File: leetcode7/leetcode_7_1.cpp
class Solution {
public:
    int reverse(int x) {
        if(x == 0){
            return x;
        }
        else if( x > 0){//positive
            int output = 0;
            
            do{
                if(output > (INT_MAX / 10)){
                    return 0;
                }
                else if(output == INT_MAX / 10 && x % 10 > INT_MAX % 10){
                    return 0;
                }
                
                output *= 10;
                output += x % 10;
                x /= 10;
            }while(x != 0);
            
            return output;
        }
        else{//negative
            int output = 0;
            
            do{
                if(output < (INT_MIN / 10)){
                    return 0;
                }
                else if(output == INT_MIN / 10 && x % 10 < INT_MIN % 10){
                    return 0;
                }
                
                output *= 10;
                output += x % 10;
                x /= 10;
                
            }while(x != 0);
            
            return output;
        }
        
    }
};
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Reverse Integer.
// Memory Usage: 8.1 MB, less than 66.13% of C++ online submissions for Reverse Integer.
// Time Complexity: O(log(x)). There are roughly log (x) digits in x.
// Space Complexity: O(1).

File: leetcode7/leetcode7_1.java
// Solution 1
// Runtime: 1 ms, faster than 100.00% of Java online submissions for Reverse Integer.
// Memory Usage: 36.3 MB, less than 36.79% of Java online submissions for Reverse Integer.
// time complexity: O(log n)
// space complexity: O(1)
class Solution {
    public int reverse(int x) {
        int rev = 0;
        while (x != 0) {
            int pop = x % 10;
            x /= 10;
            if (rev > Integer.MAX_VALUE/10 || (rev == Integer.MAX_VALUE / 10 && pop > 7)) return 0;
            if (rev < Integer.MIN_VALUE/10 || (rev == Integer.MIN_VALUE / 10 && pop < -8)) return 0;
            rev = rev * 10 + pop;
        }
        return rev;
    }
}
